[WPE][Skia] Enable MSAA when available

magomez opened the pull request:
#### 1bd3d2f0f7e6380b6d9e2e6003dfb2636b51b266
<pre>
[WPE][Skia] Enable MSAA when available
<a href="https://bugs.webkit.org/show_bug.cgi?id=277800">https://bugs.webkit.org/show_bug.cgi?id=277800</a>

Reviewed by Carlos Garcia Campos.

Enable MSAA on WPE whenever possible. On x86 or x86_64, use eight samples
to get an antialiasing result similar to non MSAA. On embedded devices, use
four samples instead, as this is a nice compromise between quality and
resource usage. On GTK keep it disabled by default. The user can overwrite
this default value through the WEBKIT_MSAA_SAMPLE_COUNT env var.

Even if MSAA is reported as available by GLES, Skia has a lot of internal checks
that may disable it, so the best way to check whether we can use it with Skia is
trying to create a texture backed SkSurface with a sample count of four and
check whether it works.

* Source/WebCore/platform/graphics/PlatformDisplay.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::create):
(WebCore::flushSurfaceIfNeeded):
(WebCore::ImageBufferSkiaAcceleratedBackend::copyNativeImage):
(WebCore::ImageBufferSkiaAcceleratedBackend::createNativeImageReference):
* Source/WebCore/platform/graphics/skia/PlatformDisplaySkia.cpp:
(WebCore::initializeMSAASampleCount):
(WebCore::SkiaGLContext::sampleCount):
(WebCore::SkiaGLContext::SkiaGLContext):
(WebCore::PlatformDisplay::msaaSampleCount):
* Source/WebCore/platform/graphics/skia/SkiaAcceleratedBufferPool.cpp:
(WebCore::SkiaAcceleratedBufferPool::createAcceleratedBuffer):

Canonical link: <a href="https://commits.webkit.org/282223@main">https://commits.webkit.org/282223@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f949da483fbcafe7f97996a8873f94111de36936

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62348 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41703 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14940 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66328 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12893 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64468 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49389 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13233 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50253 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8904 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65417 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38748 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54001 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31016 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35440 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11302 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11824 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57104 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11617 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68058 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6291 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11346 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57626 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6318 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54017 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57836 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5228 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9405 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37503 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38586 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39682 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38332 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/PlatformDisplaySkia.cpp:
```diff
@@ -66,6 +66,33 @@ static sk_sp<const GrGLInterface> skiaGLInterface()
+    static int sampleCount = 0;
+
+    std::call_once(onceFlag, [grContext] {
+        const char* disableMSAA = getenv("WPE_DISABLE_MSAA");
```
I would rather make the sample count configurable, aka WEBKIT_MSAA_SAMPLE_COUNT. If unset, on WPE, assume 4. On Gtk, if unset assume 0.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/PlatformDisplaySkia.cpp:
```diff
@@ -128,6 +162,7 @@ class SkiaGLContext : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Ski
     std::unique_ptr<GLContext> m_skiaGLContext WTF_GUARDED_BY_LOCK(m_lock);
     sk_sp<GrDirectContext> m_skiaGrContext WTF_GUARDED_BY_LOCK(m_lock);
     mutable Lock m_lock;
+    int m_sampleCount { 0 };
```
IMHO we should use unsigned consistently - I am aware that Skia uses int for the sampleCount parameter, but negative sample counts don't make any sense. So I think we should choose an unsigned.

nikolaszimmermann commented on Source/WebCore/platform/graphics/PlatformDisplay.h:
```diff
@@ -145,6 +145,7 @@ class PlatformDisplay {
 #if USE(SKIA)
     GLContext* skiaGLContext();
     GrDirectContext* skiaGrContext();
+    int MSAASampleCount();
```
msaaSampleCount?

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/PlatformDisplaySkia.cpp:
```diff
@@ -66,6 +66,33 @@ static sk_sp<const GrGLInterface> skiaGLInterface()
+{
+#if PLATFORM(GTK)
+    // GTK is desktop oriented, where using the GPU for antialiasing is not that necessary.
+    return 0;
```
I think we should still check if WEBKIT_MSAA_SAMPLE_COUNT (see next comment) is set != 0, and if so use it. If unset return zero.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/PlatformDisplaySkia.cpp:
```diff
@@ -152,6 +187,11 @@ GrDirectContext* PlatformDisplay::skiaGrContext()
     return s_skiaGLContext->skiaGrContext();
 }
 
+int PlatformDisplay::MSAASampleCount()
```
I think we should make sure that this information is included in webkit://gpu.

magomez replied:
I haven't added this cause it means adding skia to the UI process, which I'm not sure it's a good idea.

carlosgcampos replied:
We are already using skia in the ui process, but I think we could add any skia info to webkit://gpu in a different patch.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/PlatformDisplaySkia.cpp:
```diff
@@ -66,6 +84,32 @@ static sk_sp<const GrGLInterface> skiaGLInterface()
+        if (!envString.isEmpty())
+            sampleCount = parseInteger<unsigned>(envString).value_or(0);
+
+        // Skia checks internally whether MSAA is supported, but also disables it for several platforms where it
+        // knows there are bugs. The only way to know whether our sample count will work is trying to create a
+        // surface with that value and check whether it works.
+        auto imageInfo = SkImageInfo::Make(512, 512, kRGBA_8888_SkColorType, kPremul_SkAlphaType, SkColorSpace::MakeSRGB());
+        SkSurfaceProps properties = { 0, FontRenderOptions::singleton().subpixelOrder() };
+        auto surface = SkSurfaces::RenderTarget(grContext, skgpu::Budgeted::kNo, imageInfo, sampleCount, kTopLeft_GrSurfaceOrigin, &properties);
```
If the env var set the value to 0 or 1, we don't need to do this check

carlosgcampos commented on Source/WebCore/platform/graphics/skia/PlatformDisplaySkia.cpp:
```diff
@@ -66,6 +84,32 @@ static sk_sp<const GrGLInterface> skiaGLInterface()
+
+    std::call_once(onceFlag, [grContext] {
+        // Let the user override the default sample count if they want to.
+        String envString = String::fromLatin1(getenv("WEBKIT_MSAA_SAMPLE_COUNT"));
```
```suggestion
        String envString = String::fromLatin1(getenv("WEBKIT_SKIA_MSAA_SAMPLE_COUNT"));
```

carlosgcampos commented on Source/WebCore/platform/graphics/skia/PlatformDisplaySkia.cpp:
```diff
@@ -102,6 +146,11 @@ class SkiaGLContext : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Ski
         return m_skiaGrContext.get();
     }
 
+    unsigned sampleCount()
```
```suggestion
    unsigned sampleCount() const
```

carlosgcampos commented on Source/WebCore/platform/graphics/skia/PlatformDisplaySkia.cpp:
```diff
@@ -115,6 +164,8 @@ class SkiaGLContext : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Ski
             m_skiaGLContext = WTFMove(glContext);
             m_skiaGrContext = WTFMove(grContext);
         }
+
+        m_sampleCount = initializeMSAASampleCount(m_skiaGrContext.get());
```
Move this to the previous if, because MakeGL can fail, in which case m_skiaGrContext is nullptr.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/PlatformDisplaySkia.cpp:
```diff
@@ -152,6 +204,11 @@ GrDirectContext* PlatformDisplay::skiaGrContext()
     return s_skiaGLContext->skiaGrContext();
 }
 
+unsigned PlatformDisplay::msaaSampleCount()
```
```suggestion
unsigned PlatformDisplay::msaaSampleCount() const
```

magomez committed:
[WPE][Skia] Enable MSAA when available
https://bugs.webkit.org/show_bug.cgi?id=277800

Reviewed by Carlos Garcia Campos.

Enable MSAA on WPE whenever possible. On x86 or x86_64, use eight samples
to get an antialiasing result similar to non MSAA. On embedded devices, use
four samples instead, as this is a nice compromise between quality and
resource usage. On GTK keep it disabled by default. The user can overwrite
this default value through the WEBKIT_MSAA_SAMPLE_COUNT env var.

Even if MSAA is reported as available by GLES, Skia has a lot of internal checks
that may disable it, so the best way to check whether we can use it with Skia is
trying to create a texture backed SkSurface with a sample count of four and
check whether it works.

* Source/WebCore/platform/graphics/PlatformDisplay.h:
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::create):
(WebCore::flushSurfaceIfNeeded):
(WebCore::ImageBufferSkiaAcceleratedBackend::copyNativeImage):
(WebCore::ImageBufferSkiaAcceleratedBackend::createNativeImageReference):
* Source/WebCore/platform/graphics/skia/PlatformDisplaySkia.cpp:
(WebCore::initializeMSAASampleCount):
(WebCore::SkiaGLContext::sampleCount):
(WebCore::SkiaGLContext::SkiaGLContext):
(WebCore::PlatformDisplay::msaaSampleCount):
* Source/WebCore/platform/graphics/skia/SkiaAcceleratedBufferPool.cpp:
(WebCore::SkiaAcceleratedBufferPool::createAcceleratedBuffer):

Canonical link: https://commits.webkit.org/282223@main
